infrasim/model/elements/pcie_topology.py

```diff
@@ -0,0 +1,111 @@
+'''
+*********************************************************
+Copyright @ 2015 EMC Corporation All Rights Reserved
+*********************************************************
+'''
+# -*- coding: utf-8 -*-
+
+
+from infrasim.model.core.element import CElement
+from pcie_rootport import CPCIERootport
+from pcie_upstream import CPCIEUpstream
+from pcie_downstream import CPCIEDownstream
+from infrasim.helper import fw_cfg_file_create
+from infrasim import ArgsNotCorrect
+
+
+class CPCIETopology(CElement):
+    def __init__(self, pcie_topology):
+        super(CPCIETopology, self).__init__()
+        self.__pcie_topology = pcie_topology
+        self.__pcie_option = None
+        self.__workspace = None
+        self.fw_cfg_file = None
+        self.__component_list = []
+
+    def set_own_option(self, option, position):
+        if self.__pcie_option is None:
+            self.__pcie_option = option
+            return
+
+        if position is True:
+            self.__pcie_option = " ".join(["{}".format(self.__pcie_option),
+                                         "{}".format(option)])
+        else:
+            self.__pcie_option = " ".join(["{}".format(option),
+                                         "{}".format(self.__pcie_option)])
+
+    def build_topo(self, component):
+        list_tmp = []
+        for component_tmp in self.__component_list:
+            if component_tmp["bus"] == component["id"]:
+                list_tmp.append(component_tmp)
+
+        if len(list_tmp):
+            for component_tmp in list_tmp:
+                self.build_topo(component_tmp)
+
+        self.set_own_option(component["option"], False)
+        return
+
+    def set_workspace(self, ws):
+        self.__workspace = ws
+
+    def get_workspace(self):
+        return self.__workspace
+
+    def precheck(self):
+        if self.__pcie_topology is None:
+            self.logger.exception("[PCIETopology] PCIE topology is required.")
+            raise ArgsNotCorrect("pci topology is required.")
```

Comment:
PCI bus number should be uniq globally, and `CPCIETopology.precheck()` is a good place to do this check, I think.